pkg/memtier/tracker_finder.go

```diff
@@ -0,0 +1,269 @@
+	regionsMutex sync.Mutex
+	config       *TrackerFinderConfig
+	regions      map[int]*AddrRanges
+	toSampler    chan byte
+}
+
+func init() {
+	TrackerRegister("finder", NewTrackerFinder)
+}
+
+// NewTrackerFinder creates a new instance of TrackerFinder.
+func NewTrackerFinder() (Tracker, error) {
+	t := &TrackerFinder{
+		regions: make(map[int]*AddrRanges),
+	}
+	err := t.SetConfigJSON(trackerFinderDefaults)
+	if err != nil {
+		return nil, fmt.Errorf("invalid finder default configuration")
+	}
+	return t, nil
+}
+
+// SetConfigJSON sets the configuration for TrackerFinder from a JSON string.
+func (t *TrackerFinder) SetConfigJSON(configJSON string) error {
+	config := &TrackerFinderConfig{}
+	if err := unmarshal(configJSON, config); err != nil {
+		return err
+	}
+	t.config = config
+	return nil
+}
+
+// GetConfigJSON returns the JSON representation of the TrackerFinder's configuration.
+func (t *TrackerFinder) GetConfigJSON() string {
+	if t.config == nil {
+		return ""
+	}
+	if configStr, err := json.Marshal(t.config); err == nil {
+		return string(configStr)
+	}
+	return ""
+}
+
+func (t *TrackerFinder) addRanges(pid int) {
+	delete(t.regions, pid)
+	p := NewProcess(pid)
+	if ars, err := p.AddressRanges(); err == nil {
+		t.regions[pid] = ars
+	}
+}
+
+// AddPids adds PIDs to the TrackerFinder's regions map.
+func (t *TrackerFinder) AddPids(pids []int) {
+	log.Debugf("TrackerFinder: AddPids(%v)\n", pids)
+	for _, pid := range pids {
+		t.regionsMutex.Lock()
+		t.addRanges(pid)
+		t.regionsMutex.Unlock()
+	}
+}
+
+// RemovePids removes PIDs from the TrackerFinder's regions map.
+func (t *TrackerFinder) RemovePids(pids []int) {
+	log.Debugf("TrackerFinder: RemovePids(%v)\n", pids)
+	if pids == nil {
+		t.regionsMutex.Lock()
+		t.regions = make(map[int]*AddrRanges, 0)
+		t.regionsMutex.Unlock()
+		return
+	}
+	for _, pid := range pids {
+		t.removePid(pid)
+	}
+}
+
+func (t *TrackerFinder) removePid(pid int) {
+	t.regionsMutex.Lock()
+	delete(t.regions, pid)
+	t.regionsMutex.Unlock()
+}
+
+// ResetCounters does nothing as TrackerFinder tracks no memory accesses.
+func (t *TrackerFinder) ResetCounters() {
+}
+
+// GetCounters returns the counters tracked by TrackerFinder.
+func (t *TrackerFinder) GetCounters() *TrackerCounters {
+	t.regionsMutex.Lock()
+	defer t.regionsMutex.Unlock()
+	pageAttributes := uint64(0)
+	if t.config.PagemapBitPresent != nil {
+		if *t.config.PagemapBitPresent {
+			pageAttributes |= PMPresentSet
+		} else {
+			pageAttributes |= PMPresentCleared
+		}
+	}
+	if t.config.PagemapBitExclusive != nil {
+		if *t.config.PagemapBitExclusive {
+			pageAttributes |= PMExclusiveSet
+		} else {
+			pageAttributes |= PMExclusiveCleared
+		}
+	}
+	if t.config.PagemapBitSoftDirty != nil {
+		if *t.config.PagemapBitSoftDirty {
+			pageAttributes |= PMDirtySet
+		} else {
+			pageAttributes |= PMDirtyCleared
+		}
+	}
+	tcs := &TrackerCounters{}
+	for _, addrRanges := range t.regions {
+		ars := addrRanges
+		if (pageAttributes != 0) {
```

Comment:
Just use "if pageAttributes != 0" instead of "if (pageAttributes != 0)"?
It is revealed by "make format", there are other formatting issues left, maybe running "make format" before pushing?